### PR TITLE
[Core, mod_conference] Fixed oscillation in conference AGC processing. Especially during volume reduction processing

### PR DESCRIPTION
**Pull request by figaro2015**
https://github.com/signalwire/freeswitch/issues/1011

### CODE REVIEW

**Review by crienzo — CHANGES_REQUESTED** [review]
Can this be fixed without changing switch_normalize_volume_granular()?

File: src/mod/applications/mod_conference/conference_api.c [hunk 1]
```diff
@@ -1276,15 +1276,6 @@ void conference_api_set_agc(conference_member_t *member, const char *data)
...
 
-	if (argv[0]) {
-		tmp = atoi(argv[0]);
-
-		if (tmp > 0) {
-			member->agc_period_len = (1000 / member->conference->interval) * tmp;
-		}
-	}
-
-
```
Comment (crienzo) [inline src/mod/applications/mod_conference/conference_api.c:1287]: This should be if (argv[3]) {

File: src/mod/applications/mod_conference/conference_api.c [hunk 2]
```diff
@@ -1276,15 +1276,6 @@ void conference_api_set_agc(conference_member_t *member, const char *data)
 	}
 
 
-	if (argv[0]) {
```
Comment (crienzo) [inline src/mod/applications/mod_conference/conference_api.c:1279]: if (argv[3]) {

File: src/mod/applications/mod_conference/conference_api.c [hunk 3]
```diff
@@ -1276,15 +1276,6 @@ void conference_api_set_agc(conference_member_t *member, const char *data)
...
-		tmp = atoi(argv[0]);
```
Comment (crienzo) [inline src/mod/applications/mod_conference/conference_api.c:1280]: tmp = atoi(argv[3]);

**Comment by figaro2015** [comment]
Actually, my main change is not in the switch_normalize_volume_granular(), but in switch_change_sln_volume_granular().
The main reason for the oscillation is the fact that the step in the array of coefficients for the negative adjustment is too wide. In other words, it is not 'granular' enough. Which causes, on every run of the algorithm, to force the adjustment either to increase or decrease the volume. But the actual AGC processing algorithm is correct is working fine.
Also, I had to increase the margin value to 100. 
These adjustments seem to produce acceptable results in AGC processing. To accommodate even more fine margin (20, etc) the array of negative coefficients will have to be increased even further, with the smaller step between the elements.

**Comment by crienzo** [comment]
How about calculating amplitude gain from dB like:
```
#define dB_gain_factor(gain_dB) (powf(10.0f, (float)gain_dB / 20.0f))
```
Then we get 2x perceived volume increase for every +10 increase in gain_dB.  Then you can be as granular as you want to be.

**Comment by crienzo** [comment]
@figaro2015 I rescaled the tables for the granular volume.  Looks like they are still pretty close- can you verify this works for you?  Thanks.

**Comment by figaro2015** [comment]
Chris, I believe the coefficients you generated are working fine. Please see the screenshot of the Audacity plot attached.

Latest_AGC

**Comment by crienzo** [comment]
@andywolk I think we can take this now

**Review by crienzo — APPROVED** [review]

**Review by andywolk — APPROVED** [review]

**Comment by MarioG-X** [comment]
I think it's possible that this change broke building make now fails. I was able to build on March 24, but not on April 3. Using git log the only change between the two is this issue. Please see issue #1145.

git log:
commit ec1a0215fedb56bcd6f73fb1e93f763ed5286cbb (HEAD -> master, origin/master, origin/HEAD)
Author: figaro2015 <figaro2015@users.noreply.github.com>
Date:   Fri Apr 2 12:18:16 2021 -0700

    [Core, mod_conference] Fixed oscillation in conference AGC processing. Especially during volume reduction processing ( 1 0 6 4)
    * Fixed oscillation in conference AGC processing. Especially during volume reduction processing    
    * Re-scaled granular volume to +/- 50 dB level change. 10 = double loudness, -10 = half loudness, 0 = no change
    Co-authored-by: Chris Rienzo <chris@signalwire.com>
commit 67cec5c3e80e653d4540b9ba2c4a0b38b92a9458
Author: Andrey Volk <andywolk@gmail.com>
Date:   Thu Mar 25 18:52:56 2021 +0300
    version bump

**Pull request merged**